src/Microsoft.VisualStudio.Services.Agent/DistributedTask.Expressions/LexicalAnalyzer.cs

```diff
@@ -0,0 +1,318 @@
+
+            _lastToken = token;
+            return true;
+        }
+
+        private Token ReadNumberOrVersionToken()
+        {
+            int startIndex = _index;
+            int periods = 0;
+            do
+            {
+                if (_raw[_index] == '.')
+                {
+                    periods++;
+                }
+
+                _index++;
+            }
+            while (_index < _raw.Length && (!TestWhitespaceOrPunctuation(_raw[_index]) || _raw[_index] == '.'));
+
+            int length = _index - startIndex;
+            string str = _raw.Substring(startIndex, length);
+            if (periods >= 2)
+            {
+                Version version;
+                if (Version.TryParse(str, out version))
+                {
+                    return new Token(TokenKind.Version, startIndex, length, version);
+                }
+            }
+            else
+            {
+                // Note, NumberStyles.AllowThousands cannot be allowed since comma has special meaning as a token separator.
+                decimal d;
+                if (decimal.TryParse(
+                        str,
+                        NumberStyles.AllowDecimalPoint | NumberStyles.AllowLeadingSign,
+                        CultureInfo.InvariantCulture,
+                        out d))
+                {
+                    return new Token(TokenKind.Number, startIndex, length, d);
+                }
+            }
+
+            return new Token(TokenKind.Unrecognized, startIndex, length);
+        }
+
+        private Token ReadKeywordToken()
+        {
+            // Read to the end of the keyword.
+            int startIndex = _index;
+            _index++; // Skip the first char. It is already known to be the start of the keyword.
+            while (_index < _raw.Length && !TestWhitespaceOrPunctuation(_raw[_index]))
+            {
+                _index++;
+            }
+
+            // Test if valid keyword character sequence.
+            int length = _index - startIndex;
+            string str = _raw.Substring(startIndex, length);
+            if (s_keywordRegex.IsMatch(str))
+            {
+                // Test if follows property dereference operator.
+                if (_lastToken != null && _lastToken.Kind == TokenKind.Dereference)
+                {
+                    return new Token(TokenKind.PropertyName, startIndex, length);
+                }
+
+                // Boolean
+                if (str.Equals(bool.TrueString, StringComparison.OrdinalIgnoreCase))
+                {
+                    return new Token(TokenKind.Boolean, startIndex, length, true);
+                }
+                else if (str.Equals(bool.FalseString, StringComparison.OrdinalIgnoreCase))
+                {
+                    return new Token(TokenKind.Boolean, startIndex, length, false);
+                }
+                // Function
+                else if (str.Equals(And, StringComparison.OrdinalIgnoreCase))
```

Comment:
I'm thinking about switching this over to lookup well-known functions in a hashtable of function info (similar to how extension functions are registered). Right now, the IExtensionInfo contract is `string Name`, `int MinParameterCount`, `int MaxParameterCount`, and `FunctionNode CreateNode()`.

It would reduce the code here a bit (eliminate big list of if-else-if checks), and a 3 or 4 more places in parser.cs also.